### PR TITLE
Implement core functionality

### PR DESCRIPTION
**Pull request by namoscato**
- [x] cleanup message builder patterns
- [x] broadcast to channel on error
- [x] support multiple intermediate stages
- [x] unit tests
- [x] readme / documentation
- [x] status CI badge
- [x] marketplace branding
- [x] update duration algorithm
- [x] update `@dev` references

### CODE REVIEW

File: src/github/summary.ts [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+
+  return createMessage(text, contextBlock)
+}
+
+function getText(status?: string): Text {
+  const summarySentence = getSummarySentence(status)
+  const eventLink = getEventLink()
+
+  const mrkdwn = [
+    status ? emojiFromStatus(status) : emoji('black_square_button'),
```
Comment (namoscato) [inline src/github/summary.ts:40]: I don't personally love the initial / in progress / "unchecked" icon:

Screen Shot 2022-09-11 at 3 22 59 PM

Perhaps we make this configurable at some point.

File: src/github/stage.ts [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    .flatMap<CompletedJobStep>(({steps}) => {
+      if (!steps) {
+        return []
+      }
+
+      return steps
+        .filter(isCompletedJobStep)
+        .filter(({name}) => slackRegex.test(` ${name} `))
+    })
+    .pop()
```
Comment (namoscato) [inline src/github/stage.ts:89]: This won't work well with _parallel_ jobs. Something to probably revisit in https://github.com/namoscato/action-slack-deploy-pipeline/issues/12.

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+      SLACK_DEPLOY_BOT_TOKEN: ${{ secrets.SLACK_BOT_TOKEN_FOR_TESTING }}
+      SLACK_DEPLOY_CHANNEL: 'C040YVCUDRR' #github_action_testing
+    steps:
+      - uses: actions/checkout@v3
+      - uses: ./
+        id: slack
+      - uses: ./
+        with:
+          thread_ts: ${{ steps.slack.outputs.ts }}
+          conclusion: true
```
Comment (johnkoehn) [inline .github/workflows/ci.yml:33]: :D

File: src/github/stage.ts (hunk 2 above)
Comment (johnkoehn) [reply src/github/stage.ts:89]: 1. So this grab the last completed slack step, regardless of what job it is? 
2. Do you think taking in the job name as an optional parameter would solve for this?

If one is yes, I'm actually a bit surprised this works. I guess the assumption here is the jobs array returned is "ordered". I wonder however, if I have a job in the middle of my workflow but it runs last, would that break this?

File: src/github/summary.ts (hunk 1 above)
Comment (johnkoehn) [reply src/github/summary.ts:40]: That'd be sweet! Big fan of using custom emojis :D

**Review by johnkoehn — COMMENTED** [review]
This is awesome!

Left a couple questions around last completed

File: src/github/stage.ts (hunk 2 above)
Comment (namoscato) [reply src/github/stage.ts:89]: > So this grab the last completed slack step, regardless of what job it is?

Correct.

In a sequential workflow (in which jobs depend on each other), this seems to work fine; the `listJobsForWorkflowRun` response does not return jobs that have not yet started.

> Do you think taking in the job name as an optional parameter would solve for this?

Do you mean an optional input specifying the _previous_ / dependent job name? We inherently know which job we are currently in, but we actually need to know which _previous_ job / step to compute duration.

Comment (namoscato) [reply src/github/stage.ts:89]: Having given this some more thought, I _am_ leaning towards scoping "stage" duration to its contextual job. We'll still want to support multiple Slack steps per job, but I now see less value in computing this across jobs, especially given the threaded message displays the job's name.

Comment (namoscato) [reply src/github/stage.ts]: Updated in c6eb5f9bec65dd26fe211cd9766f63d888d0f246.

It still favors a `lastCompletedSlackStep` but scoped within the current job, which still allows for scenarios in which multiple slack notifications can be sent within the same job, i.e.

```yaml
steps:
- name: Post to Slack
- name: Deploy to staging
- name: Post to Slack
- name: Deploy to prod
- name: Post to Slack
```

File: src/github/summary.ts (hunk 1 above)
Comment (namoscato) [reply src/github/summary.ts:40]: Tracking via https://github.com/namoscato/action-slack-deploy-pipeline/issues/13.

**Pull request merged**